S1-Based 3D Position Reconstruction

matteoguida opened the pull request:
## What does the code in this PR do / what does it improve?
This PR adds **S1-based position reconstruction** methods at the event level. 

It was decided that implementing it at peak level was not necessary. The script is copied from the [standard position reconstruction](https://github.com/XENONnT/straxen/blob/master/straxen/plugins/position_reconstruction.py), so that different models can be managed in the future. 

At the current stage only the **CNN data-driven** (see [here](https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:xenonnt:analysis:reconstruction_team:s1_xyz_rec_with_data_driven)) is included. 

The **runtime** per event should be bounded by more than one order of magnitude by "event_area_per_channel" and it results now in ≈ 6 ms per event. 
 
Possible applications: to map possible charge insensitive regions, radon S1 only analysis, improve event quality selections in high energy.

## Can you briefly describe how it works?

Loaded the trained model, for each event the (x,y,z) position in cm is returned taking as input the PMT hit patterns of the S1 signal, after that, they have been normalized by dividing by the channel with maximum area. 

## Can you give a minimal working example (or illustrate with a figure)?
Example with more explanations: [script](https://github.com/XENONnT/analysiscode/tree/master/s1_pos_rec)

Example of prediction:

![example_prediction](https://user-images.githubusercontent.com/44644744/135132642-c07182dd-367d-47df-852e-0e7a043980d6.PNG)

dachengx commented:
Great job @matteoguida ! Would you plan to merge the master into `xedocs_URL` branch?

matteoguida commented:
@dachengx done, thanks!

SphericalCowww commented:
Hello Matteo, thank you very much for your wonderful S1 CNN position reconstruction~ I am planning to use your `event_s1_position_cnn` for s1_only analysis especially when the anode is off and so no S2 peaks. Previously, I was using the data kind `peaks` for the analysis because it finds the peaks individually. For data kind `events` to my understanding, requires an S1/S2 pair to form an event. So I am wondering what the best strategy would be in this case; the analysis needs the S1 area and positions, so 4 columns would suffice.

The `exclude_s1_as_triggering_peaks` on `events` could be a neat solution though as you have told me. The only thing I would be wondering about is then if we will need to regenerate the `events` data for this option.

Thanks in advance!

jmosbacher commented:
I agree with @dachengx, this approach makes most conceptual sense to me.

dachengx reviewed (changes requested):
I suggest we modify this PR into:
1. peak-level S1 pos-rec
2. plugin of event-level S1 pos-rec to provide the opportunity to rerun the pos-rec
Please let me know if and how we should collaborate on this modification. : )

matteoguida commented:
@dachengx I added the files for S1-posted at peak level. The ones for the event level were already implemented. I also updated [here](https://github.com/XENONnT/analysiscode/tree/master/s1_pos_rec) the with eye tests including both event and peak level cases.

@LuisSanchez25 The peak level implementation now is working using the line `tf_event_model_s1_cnn = straxen.URLConfig.evaluate_dry` in the file `straxen/straxen/plugins/peaks/peak_s1_position_cnn.py` while using `straxen.URLConfig` there is a xedocs related error. I do not know what is necessary to change in the xedocs repo, I guess/hope it is just one line of code or so..
If you can please do this modification ad see if [this ](https://github.com/XENONnT/analysiscode/blob/master/s1_pos_rec/test_s1_cnn_peak_level.py) script then is running with your modification would be great.

Please as you can see [here](https://xenonnt.slack.com/archives/C049JJZ2U3A/p1680791889984329) there is a call for updates in straxen where it would be great to include also this PR before reprocessing. In light of this if these modifications could be done today or so it would be superb, so that @dachengx can then see if we can merge

LuisSanchez25 commented:
I will take a look at it in a bit and will let you know, the URL was working from my local environment when we first talked about this so my guess is that it might be related to some versioning issue.

matteoguida commented on straxen/plugins/peaks/_peak_s1_position_base.py:
```diff
@@ -0,0 +1,94 @@
+        model = self.get_tf_model()
+
+        # Reconstruct position only for large peaks, otherwise severe inaccuracy.
+        peak_mask = (peaks['area_per_channel'].sum(axis=1) > self.config['min_s1_area_s1_posrec'])&(peaks['type']<2)
```
This is a good point, I would say there is not a well-defined motivation. In the S2 version `_peak_positions_base.py` there is no filter: the predictions are made on peaks classified as 0: unknown, 1: S1, 2: S2. 

Not putting any filters, in this case, seemed unreasonable to me. I thought: since there is already the area>1000 filter, we also make predictions on 0 in case there has been some classification difficulty which perhaps can be detected later. But we could also directly put `peaks['type']==1` if you think is better.

dachengx replied:
Let's just stay with the current code.
